sim_pipeline/halos/halos.py

```diff
@@ -0,0 +1,235 @@
+    :param m_max: The maximum halo mass.
+    :type m_max: float, optional
+    :param resolution: The resolution of the mass grid.
+    :type resolution: int, optional
+    :param wavenumber: The wavenumber array for power spectrum.
+    :type wavenumber: array, optional
+    :param collapse_function: The halo collapse function.
+    :type collapse_function: function, optional
+    :param power_spectrum: The power spectrum function.
+    :type power_spectrum: function, optional
+    :param params: The parameters for the collapse function.
+    :type params: tuple, optional
+    :return: An array of redshifts of halos.
+    :rtype: array
+    """
+    if cosmology is None:
+        warnings.warn("No cosmology provided, instead uses flat LCDM with default parameters")
+        from astropy.cosmology import FlatLambdaCDM
+        cosmology = FlatLambdaCDM(H0=70, Om0=0.3)
+
+    dN_dz = (cosmology.differential_comoving_volume(redshift_list) * sky_area).to_value('Mpc3')
+    density = number_density_at_redshift(z=redshift_list, m_min=m_min, m_max=m_max, resolution=resolution,
+                                         wavenumber=wavenumber,
+                                         power_spectrum=power_spectrum, cosmology=cosmology,
+                                         collapse_function=collapse_function, params=params)
+    dN_dz *= density
+
+    # integrate density to get expected number of galaxies
+    N = np.trapz(dN_dz, redshift_list)
+    N = int(N)
+
+    # cumulative trapezoidal rule to get redshift CDF
+    cdf = dN_dz  # reuse memory
+    np.cumsum((dN_dz[1:] + dN_dz[:-1]) / 2 * np.diff(redshift_list), out=cdf[1:])
+    cdf[0] = 0
+    cdf /= cdf[-1]
+
+    # sample N galaxy redshifts
+    return np.interp(np.random.rand(N), cdf, redshift_list)
+
+
+def halo_mass_at_z(z, m_min=None, m_max=None, resolution=None, wavenumber=None, power_spectrum=None, cosmology=None,
+                   collapse_function=None, params=None):
+    """
+    Function to calculate the mass of halos at a given redshift.
+
+    :param z: The redshift at which to evaluate the halo mass.
+    :type z: float
+    :param m_min: The minimum halo mass.
+    :type m_min: float, optional
+    :param m_max: The maximum halo mass.
+    :type m_max: float, optional
+    :param resolution: The resolution of the mass grid.
+    :type resolution: int, optional
+    :param wavenumber: The wavenumber array for power spectrum.
+    :type wavenumber: array, optional
+    :param power_spectrum: The power spectrum function.
+    :type power_spectrum: function, optional
+    :param cosmology: The cosmology instance to use.
+    :type cosmology: Cosmology, optional
+    :param collapse_function: The halo collapse function.
+    :type collapse_function: function, optional
+    :param params: The parameters for the collapse function.
+    :type params: tuple, optional
+    :return: The mass of halos at redshift z.
+    :rtype: array
+    """
+    # TODO: debug
+    if m_min is None:
```

Comment:
these default values settings (I think) you have in two routines. Perhaps you can make one 'utility' function that provides these settings and can be called for two or more functions, to avoid duplication

Reply:
@sibirrer See the new update for halos.py